---
title: "MB Sex Differences in DNA Methylation in subgroups - call CNVs"
author: "Rachel Moss"
date: "June 27, 2022"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---

## Introduction

I will use [conumee](https://bioconductor.org/packages/release/bioc/html/conumee.html) to call copy number variations (CNVs) from the Illumina 450k methylation arrays in the Newcastle and the Cavalli datasets. 
The GSE15745 healthy brain dataset does not have idat files available, so I will use the normal cerebellum samples from [GSE134379](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134379) instead for the controls for the CNV analysis.

## Setup

### Load packages

```{r packages, message=FALSE, results=FALSE, warning=FALSE}
library(minfi)
library(conumee)
library(GEOquery)
library(tidyverse)
library(readr)
library(filesstrings)
```

### Specify files and paths

These Newcastle files are stored on a shared [google drive](https://drive.google.com/drive/folders/1QakbU5kqgWX5cQIPI64F8srgZbKbhDLj?usp=sharing).
The healthy brain data will be downloaded from GEO.

```{r inputs}
# Set the directory where the processed data will be kept
rawdata <- "/Volumes/GoogleDrive/My Drive/Research/SexDiffMedullo/Analysis_RMM_new/data/raw"
procData <- "/Volumes/GoogleDrive/My Drive/Research/SexDiffMedullo/Analysis_RMM_new/data/processed"
controlGSE <- "GSE134379"
phenoNewFile <- "Newcastle_pheno_v3.csv"


# # Set the directory where the counts files are found
# counts_dir <- here::here("analyses", "mutation_counts")
# 
# # File paths to the counts files for LGAT and medulloblastoma
# lgat_path <- file.path(counts_dir, "LGAT_gene-mutations.tsv")
# mb_path <- file.path(counts_dir, "Medulloblastoma_gene-mutations.tsv")
# 
# # Reference directory & FLAGS (frequently mutated genes)
# ref_dir <- here::here("data", "reference")
# flags_path <- file.path(ref_dir, "FLAGS.txt")
```

Outputs
```{r outputs}
# set the directory to save output plots
results_dir <- "../../results/"

# create the results directory if it does not exist
if (!dir.exists(results_dir)){
  dir.create(results_dir, recursive = TRUE)
}
```

### Download Data

```{r downloads}
# #download GEO object
# gseCtrl <- getGEO(controlGSE, GSEMatrix = TRUE)
# #get phenotype data - sample sheet
# pd = pData(gseCtrl[[1]])
# pd_filt <- pd[,c("geo_accession","title","source_name_ch1","supplementary_file","supplementary_file.1","age:ch1","bead chip:ch1","brain region:ch1","diagnosis:ch1","gender:ch1","plate:ch1","tissue:ch1","well position:ch1")]
# pd_filt$filename <- sub(".*suppl/","",pd$supplementary_file)
# pd_filt$Basename <- sub("_Grn.idat.gz","",pd_filt$filename)
# pd_filt <- pd_filt %>%
#   separate(col=Basename,into=c("GSM","Sentrix_ID","Sentrix_Position"),sep="_",remove=FALSE)
# rownames(pd_filt) <- pd_filt$geo_accession
# colnames(pd_filt) <- c("Sample_Name","Title","Source_Name","supplementary_file","supplementary_file.1","Age","Bead_Chip","Brain_Region","Sample_Group","Sex","Sample_Plate","Tissue","Sample_Well","filename","Basename","GSM","Sentrix_ID","Sentrix_Position")
# pd_filt$MSA_Well <- NA
# pd_filt$Pool_ID <- NA
# pd_filt <- pd_filt[,c("Sample_Name","Sample_Well","MSA_Well","Sample_Plate","Sample_Group","Pool_ID","Sentrix_ID","Sentrix_Position","Basename","Title","Source_Name","supplementary_file","supplementary_file.1","Age","Bead_Chip","Brain_Region","Sex","Tissue","GSM","filename")]
# #write_csv(pd_filt, file=paste0(procData,"/",controlGSE,"_pheno.csv"))
# pd_normCBL <- pd_filt %>%
#   filter(Sample_Group=="ND" & Brain_Region=="CBL")
# # write_csv(pd_normCBL, file=paste0(rawdata,"/control_",controlGSE,"/idat/","Control_CBL_",controlGSE,"_pheno_Sample_Sheet.csv"))

# #decompress idats
# dir <- paste0(rawdata,"/control_",controlGSE,"/")
# untar(paste0(dir,controlGSE,"_RAW.tar"), exdir = paste0(dir,"/idat"))
# #list files
# head(list.files(paste0(dir,"/idat"), pattern = "idat"))
# idatFiles <- list.files(paste0(dir,"/idat"), pattern = "idat.gz$", full = TRUE)
# #norm_idatFiles <- paste0(dir,"idat/",pd_normCBL$filename)
# #decompress individual idat files only the CBL and non diseased samples
# sapply(idatFiles, gunzip, overwrite = TRUE)
# 
# # Find sample sheet with phenotype info, csv file
# targets <- read.metharray.sheet(rawdata, pattern = "Sample_Sheet.csv")
# 
# # Create RGChannelSet object that contains raw intensities in the green and red channels, including internal control probes
# RGSet_Ctrl <- read.metharray.exp(targets = targets, verbose = T)
# RGSet_Ctrl
# saveRDS(RGSet_Ctrl, paste0(procData,"/RGSet_control_CBL_",controlGSE,".RDS"))
# 
# # Create methylset
# Mset_Ctrl <- preprocessRaw(RGSet_Ctrl)
# saveRDS(Mset_Ctrl, paste0(procData,"/MSet_control_CBL_",controlGSE,".RDS"))
#         
# RGSet_New <- readRDS(paste0(procData,"/RGSet_GSE93646_filt.RDS"))
# 
# # Create methyl set
# MSet_new <- preprocessRaw(RGSet_New) 
# saveRDS(MSet_new, paste0(procData,"/MSet_Newcastle_GSE93646_filt.RDS"))
phenoCtrl <- read_csv(paste0(rawdata,"/control_",controlGSE,"/idat/","Control_CBL_",controlGSE,"_pheno_Sample_Sheet.csv"))
phenoCtrl <- phenoCtrl[match(colnames(MSet_Ctrl),phenoCtrl$Basename),]
```

### Define functions
No functions right now

```{r functions}
# Function that takes a project name, MSet, and RGSet and creates various QC Plots and a pdf QC Report
minfiQC <- function(proj, dir, MSet, RGSet) {
  groups = pData(RGSet)$Sample_Group
  print(length(groups))
  names = pData(RGSet)$Sample_Name
  print(length(names))

  # QC Plot
  qc <- getQC(MSet)
  plotQC(qc)

  densityPlot(MSet, sampGroups = groups)
  densityBeanPlot(MSet, sampGroups = groups)

  # Control probes plot
  controlStripPlot(RGSet, controls="BISULFITE CONVERSION II")

  # Quality Control Report
  reportName = paste("qcReport_",proj,".pdf",sep="")
  qcReport(RGSet, sampNames = names,sampGroups = groups, pdf = paste0(dir,"/",reportName))
  return(qc)
}
```

## QC of Healthy Brain data
When I loaded the control data into conumee if gave an error warning that intensities are abnormally low (<5000) so I will run basic minfi QC on it to see if there are any bad samples.
```{r qc}
proj <- paste0("control_",controlGSE)
#qcCtrl <- minfiQC(proj, results_dir, Mset_Ctrl,)


qc <- getQC(Mset_Ctrl)
png(file=paste0(results_dir,"/",proj,"_plotQC.png"))
plotQC(qc)
dev.off()
MSet_Ctrl <- addQC(Mset_Ctrl, qc)

densityPlot(MSet_Ctrl, sampGroups = phenoCtrl$Sex)
densityBeanPlot(MSet_Ctrl, sampGroups = phenoCtrl$Sex)

# Quality Control Report
qcReport(RGset_Ctrl, sampNames = phenoCtrl$Basename,sampGroups = phenoCtrl$Sex, pdf = paste0(results_dir,"/qcReport_",proj,".pdf"))
```

## Use Conumee to call CNVs for Newcastle

Now that we have done all of our setup, we can move on to the CNV analysis using conumee

### Reading data

We'll first read in the data for Newcastle and control data as methylsets. I'll also read in the pheno data for Newcastle that contains subgroups

```{r read_data}
MSet_New <- readRDS(paste0(procData,"/MSet_Newcastle_GSE93646_filt.RDS"))
phenoNew <- pheno <- read_csv(paste0(procData,"/Newcastle/",phenoNewFile))
Mset_Ctrl <- readRDS(paste0(procData,"/MSet_control_CBL_GSE134379.RDS"))
RGset_Ctrl <- readRDS(paste0(procData,"/RGSet_control_CBL_GSE134379.RDS"))
```

### Quick stats

Describe the data we have.

```{r}
cat("Newcastle data contains:", ncol(MSet_New),"medulloblastoma samples")
cat("\n") # print return between
cat("Counts of Newcastle samples by subgroup: ")
table(phenoNew$subgroup)
cat("Control data contains:", ncol(Mset_Ctrl), "samples")
```

### Create annotation object

```{r annot}
data(exclude_regions)
data(detail_regions)
head(detail_regions, n = 2)
anno <- CNV.create_anno(array_type = "450k", exclude_regions = exclude_regions, detail_regions = detail_regions)
anno
```

### Combine intensity values

```{r combineintensity}
#data(tcgaBRCA.sentrix2name)  # TCGA sample IDs are supplied with the conumee package
#sampleNames(MsetTCGA) <- tcgaBRCA.sentrix2name[sampleNames(MsetTCGA)]
Newcastle.data <- CNV.load(MSet_New)
controls.data <- CNV.load(Mset_Ctrl)
names(Newcastle.data)
## [1] "TCGA-AR-A1AU-01A-11D-A12R-05" "TCGA-AR-A1AY-01A-21D-A12R-05"
Newcastle.data
## CNV data object
##    created   : 
##   @intensity : available (2 samples, 485512 probes)

# minfi.data <- CNV.load(MsetEx)
# minfi.controls <- pData(MsetEx)$status == "normal"
# controls.data <- CNV.load(MsetControls)
```
### Main CNV Analysis

```{r cnv}
x <- CNV.fit(Newcastle.data["GSM2459851_9761749073_R04C01"], controls.data, anno)
# y <- CNV.fit(tcga.data["TCGA-AR-A1AU-01A-11D-A12R-05"], controls.data, anno)  # CopyNumber450kData package is deprecated
# z <- CNV.fit(tcga.data["TCGA-AR-A1AY-01A-21D-A12R-05"], controls.data, anno)
# y <- CNV.fit(tcga.data["TCGA-AR-A1AU-01A-11D-A12R-05"], minfi.data[minfi.controls], anno)  # minfiData control samples
# z <- CNV.fit(tcga.data["TCGA-AR-A1AY-01A-21D-A12R-05"], minfi.data[minfi.controls], anno)

x <- CNV.bin(x)
x <- CNV.detail(x)
x <- CNV.segment(x)

# y <- CNV.segment(CNV.detail(CNV.bin(y)))
# z <- CNV.segment(CNV.detail(CNV.bin(z)))

x

```

### Output plots and text files

```{r output}
#pdf("~/conumee_analysis/CNVplots.pdf", height = 9, width = 18)
CNV.genomeplot(x)
CNV.genomeplot(x, chr = "chr6")
CNV.genomeplot(x, chr = "chr10")
CNV.detailplot(x, name = "PTEN")
CNV.detailplot_wrap(x)
#dev.off()

head(CNV.write(x, what = "segments"), n = 5)


```

### Saving plots

```{r save_plots}
# ggsave(
#   file.path(plots_dir, "LGAT_mutation-counts.png"), 
#   lgat_plot,
#   width = 7, 
#   height = 7
# )
# ggsave(
#   file.path(plots_dir, "Medulloblastoma_mutation-counts.png"), 
#   mb_plot,
#   width = 7,
#   height = 7
# )
```


## Session information

We like to end all of our notebooks with a call to a function that provides a brief report about the versions of R and all packages that were used in the notebook.
This can be very useful for debugging down the line, as different versions of packages, R, or even the operating system can sometimes result in slightly or significantly differen outputs.

The most common such function is `sessionInfo()`, which is part of base R and is what we are using here, but a nice alternative is the function `sessioninfo::session_info()` (part of the `sessioninfo` package), which provides a somewhat more neatly-formatted report with a bit more information.

```{r session_info}
sessionInfo()
```
