---
title: "MB Sex Differences in DNA Methylation in subgroups - call CNVs"
author: Rachel Moss
date: June 27, 2022
output: 
  html_notebook:
    toc: true
    toc_float: true
---

## Introduction

I will use [conumee](https://bioconductor.org/packages/release/bioc/html/conumee.html) to call copy number variations (CNVs) from the Illumina 450k methylation arrays in the Newcastle and the Cavalli datasets. 
The GSE15745 healthy brain dataset does not have idat files available, so I will use the normal cerebellum samples from [GSE134379](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134379) instead for the controls for the CNV analysis.

## Setup

### Load packages

```{r packages, message=FALSE, results=FALSE, warning=FALSE}
library(minfi)
library(conumee)
library(GEOquery)
```

### Specify files and paths

These Newcastle files are stored on a shared [google drive](https://drive.google.com/drive/folders/1QakbU5kqgWX5cQIPI64F8srgZbKbhDLj?usp=sharing).
The healthy brain data will be downloaded from GEO.

```{r inputs}
# Set the directory where the processed data will be kept
rdata <- "../../data/raw"
procData <- "../../data/processed"
control <- 

# Set the directory where the counts files are found
counts_dir <- here::here("analyses", "mutation_counts")

# File paths to the counts files for LGAT and medulloblastoma
lgat_path <- file.path(counts_dir, "LGAT_gene-mutations.tsv")
mb_path <- file.path(counts_dir, "Medulloblastoma_gene-mutations.tsv")

# Reference directory & FLAGS (frequently mutated genes)
ref_dir <- here::here("data", "reference")
flags_path <- file.path(ref_dir, "FLAGS.txt")
```

Outputs
```{r outputs}
# set the directory to save output plots
results_dir <- "./results"

# create the results directory if it does not exist
if (!dir.exists(results_dir)){
  dir.create(results_dir, recursive = TRUE)
}
```

### Download Data

```{r downloads}
#increase file download timeout
options(timeout = 800)
gc()

#download GEO object
gseCtrl <- getGEO(controlGSE, GSEMatrix = TRUE)
#get phenotype data - sample sheet
pd = pData(gseCtrl[[1]])
pd <- pd[,c("geo_accession","title","source_name_ch1","supplementary_file","supplementary_file.1","age:ch1","bead chip:ch1","brain region:ch1","diagnosis:ch1","gender:ch1","plate:ch1","tissue:ch1","well position:ch1")]
pd$sampleName <- sub(".*suppl/","",pd$supplementary_file)
pd$sampleName <- sub("_Grn.idat.gz","",pd$sampleName)
rownames(pd) <- pd$sampleName
pd <- as(pd,"DataFrame")

#get raw data - idats, processed beta matrix, etc.
getGEOSuppFiles(controlGSE, baseDir = data)

#decompress idats
untar(paste0(controlGSE,"/",controlGSE,"_RAW.tar"), exdir = paste0(data,"/",controlGSE,"/idat"))
#list files
head(list.files(paste0(controlGSE,"/idat"), pattern = "idat"))
idatFiles <- list.files(paste0(controlGSE,"/idat"), pattern = "idat.gz$", full = TRUE)
#decompress individual idat files
sapply(idatFiles, gunzip, overwrite = TRUE)
#read idats and create RGSet
RGSet_Ctrl <- read.metharray.exp("GSE93646/idat", verbose=T)
pData(RGSet_Ctrl) <- pd

saveRDS(RGSet_Ctrl, paste0("RGSet_Newcastle_",controlGSE,".RDS"))
        
# Create methylset


# Download previously processed RGSet for Newcastle data
drive_download("RGSet_GSE93646_filt.RDS", path="../../data/processed/RGSet_GSE93646_filt.RDS", overwrite=TRUE)

RGSet_New <- readRDS('../../data/processed/RGSet_GSE93646_filt.RDS')

# Create methyl set
MSet_new <- preprocessRaw(RGSet) 


```

### Define functions

In this section we define any functions that we will use later in the script.
While we could define functions later, it is nice to have them grouped together near the top of the notebook for at least a few reasons: - The first is that we will always know where to look for function definitions.
Patterns and conventions make it easier for us to know what to expect when we come back to a file after a while away and may need to make changes - Another reason is that if the function definitions are at the top, we can use them anywhere below, without worrying about invoking a function before it is defined.
Finally, having the functions grouped together will make it easier if we decide to move them to an external file: - If the set of functions gets large, or if we want to use some functions across multiple notebooks, we can create a separate `.R` file (or files) and then use the `source()` command to load those functions into this notebook (or others) all at once.

In this case, we will only have one function, which will make a plot of the number of samples with mutations for each gene.
We'll give it some options to allow filtering to a minimum number of mutated samples and highlighting particular genes of interest.

Importantly, we will be sure to include some comments with documentation of the arguments for the function, as well as a description of what the function will return.

To make these comments a bit more structured, we will use the format from the [`roxygen2`](https://roxygen2.r-lib.org/) package for these comments, which would be useful for automatically generating documentation if we were creating an R package.
We aren't generating separate documentation files here, but it is still a nice enough convention, and RStudio has an option in the `Code` menu called `Insert Roxygen Skeleton` that can save us a bunch of typing and create the outline of what we need in a single step.

```{r mutation_plot_function}
#' Plot the number of  mutations for each gene from a data frame
#'  
#' Filters to a cutoff and sorts genes from most to least mutated, then 
#' creates a bar plot of the mutation counts, optionally highlighting 
#' genes of interest.
#'
#' @param mutations_df A data frame with columns `Hugo_Symbol` and `mutated_samples`
#' @param min_mutated The minimum `mutated_samples` value to include in the plot
#'  (default: 3)
#' @param highlight_genes A vector of genes to highlight in the plot (optional)
#' @param highlight_title The title for the highlighted genes legend 
#'  (default: "Gene of interest")
#'
#' @return A ggplot2 plot object
#'
plot_gene_mutations <- function(
  mutations_df,  
  min_mutated = 3,
  highlight_genes = c(),
  highlight_title = "Gene of interest"
){
  plot_df <- mutations_df %>%
    filter(mutated_samples >= min_mutated) %>%
    arrange(desc(mutated_samples)) %>%
    # make the gene names a factor, in the order they appear after sorting
    mutate(gene = factor(Hugo_Symbol, levels = Hugo_Symbol))
  
  if (length(highlight_genes) == 0){
    # If there are no highlight genes, we don't include that in the ggplot aesthetics
    plot_obj <- ggplot(plot_df, 
                       aes(x = gene, 
                           y = mutated_samples))
  } else {
    # There are highlight genes, so add them to the data frame
    plot_df <- plot_df %>%
      mutate(
        # add a column with the highlight status, make it a factor for ordering
        highlight = ifelse(gene %in% highlight_genes, "Yes", "No"),
        highlight = factor(highlight, levels = c("Yes", "No"))
      )
    # create the plot object with colored fill
    plot_obj <- ggplot(plot_df, 
                       aes(x = gene, 
                           y = mutated_samples,
                           fill = highlight)) 
  }
  
  # Add layers and formatting to the plot
  plot_obj <- plot_obj +
    geom_col() + 
    labs(x = "Gene symbol",
         y = "Mutated samples",
         fill = highlight_title) + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle=45,hjust=1)) + 
    scale_fill_brewer(palette = "Set1")
    
  # Return the plot
  return(plot_obj)
}
```

## Gene mutation exploration

Now that we have done all of our setup, we can move on to the bulk of the work.

### Reading data

We'll first read in the data for the two cancer types that we had defined input paths for in the setup section.

```{r read_data}
# read in mutation count files for plotting
lgat_df <- readr::read_tsv(lgat_path)
mb_df <- readr::read_tsv(mb_path)
```

We will also read in the table of genes frequently mutated in exons:

```{r read_flags}
flags_df <- readr::read_tsv(flags_path)
```

### Quick stats

Usually at this stage of an analysis we might want to get some quick statistics about the data that we just read in.
Here we will calculate and print out the number of genes in each of the tables.

```{r}
cat("LGAT mutated gene count:", nrow(lgat_df))
cat("\n") # print return between
cat("Medulloblastoma mutated gene count: ", nrow(mb_df))
```

**A note on code chunk names:** 
You might have noticed that in most of the previous code chunks, I included some text in the `{r}` brackets on the first line of the chunk, which serves as the name of the chunk.
The use of these names is a bit controversial among members of the Data Lab.
Some of us like them for the clarity and the fact that they allow easy navigation in RStudio.
Others find that they get out of sync with the code inside them and can be more trouble than they are worth.
In particular, no two chunks can have the same name, so sometimes duplicating or splitting a chunk can accidentally create a notebook that won't render properly.
Here we are using names for chunks with defined purpose, but not for the more exploratory, iterative chunks in this section and the next, but your preferred method is up to you!


### Make plots

We'll start by using the most basic version of the plotting function we defined earlier, with no highlighting, using the LGAT data:

```{r}
plot_gene_mutations(lgat_df)
```

We do want to make sure we highlight any frequently mutated genes, so let's add that argument, highlighting the genes from the FLAGS data set (stored in `flags_df`):

```{r}
plot_gene_mutations(lgat_df, 
                    highlight_genes = flags_df$gene, 
                    highlight_title = "Frequently mutated")
```

If we want, we can save the plot object we made into its own variable, and since it is a ggplot object, we can also add additional content or formatting that the function might not have supported.

```{r}
lgat_plot <- plot_gene_mutations(lgat_df, 
                                 highlight_genes = flags_df$gene, 
                                 highlight_title = "Frequently mutated")

# add a title to the plot
lgat_plot <- lgat_plot + labs(title = "LGAT mutation counts")

# view the plot
lgat_plot
```

Now let's repeat that for the medulloblastoma data.

```{r}
mb_plot <- plot_gene_mutations(mb_df, 
                               highlight_genes = flags_df$gene, 
                               highlight_title = "Frequently mutated")

# add a title to the plot
mb_plot <- mb_plot + labs(title = "Medulloblastoma mutation counts")

# view the plot
mb_plot
```

That plot seems to be a bit cluttered, so we will make the x axis labels a bit smaller:

```{r}
mb_plot <- mb_plot + theme(axis.text.x = element_text(size = 7))

mb_plot
```

### Saving plots

If we are happy with the plots, we can save them.
We do want to be sure to specify the width and height of the saved plots, as that can sometimes change unpredictably depending on the R environment.

We didn't define the output plot file names at the top, just the directory, so we will create the file names now (using `file.path()` again!).

In most cases, defining outputs here in the notebook will be just fine, but if the output here were likely to be feeding into another analysis, I would probably want to be more careful and define all the outputs at the top of the notebook.

```{r save_plots}
ggsave(
  file.path(plots_dir, "LGAT_mutation-counts.png"), 
  lgat_plot,
  width = 7, 
  height = 7
)
ggsave(
  file.path(plots_dir, "Medulloblastoma_mutation-counts.png"), 
  mb_plot,
  width = 7,
  height = 7
)
```


## Session information

We like to end all of our notebooks with a call to a function that provides a brief report about the versions of R and all packages that were used in the notebook.
This can be very useful for debugging down the line, as different versions of packages, R, or even the operating system can sometimes result in slightly or significantly differen outputs.

The most common such function is `sessionInfo()`, which is part of base R and is what we are using here, but a nice alternative is the function `sessioninfo::session_info()` (part of the `sessioninfo` package), which provides a somewhat more neatly-formatted report with a bit more information.

```{r session_info}
sessionInfo()
```
