---
title: "MB Sex Differences in DNA Methylation in subgroups - CNV boxplots by sex"
author: "Rachel Moss"
date: "Aug 29, 2022"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---

## Introduction

I will use UpSetR package to compare the Cavalli and Newcastle DMPs for each subgroup to the DMPs in healthy brain from  [GSE15745](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE15745)
in an upset plot. 

## Setup

### Load packages

```{r packages, message=FALSE, results=FALSE, warning=FALSE}
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(tidyverse)
library(readr)
library(ggplot2)
library(ggpubr)
library(rstatix)
```

```{r herewego}
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
ann450k_df <- as.data.frame(ann450k) %>%
  rownames_to_column(var="probeID")


```

### Specify files and paths

These DMP files for each subgroup for Cavalli and Newcastle are stored on a shared [google drive](https://drive.google.com/drive/folders/1MtKqg-v7USwvqbHVqPo_RiIwcu5O9nkb?usp=sharing).
The healthy brain DMPs are also on a shared [google drive](https://drive.google.com/drive/folders/1so_iPCp3Q4MlN-iAuQtimCt5-IzDYpDs?usp=sharing)

```{r inputs}
# Set the directory where the processed data will be kept
dir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results"
cavDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Cavalli"
newDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Newcastle"
phenoDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/data/"

# File paths to the DMP files for each subgroup
cavALL_path <- file.path(cavDMPDir, "Cavalli_agecatadj_dmp_MethFullResults.csv")
cavSHH_path <- file.path(cavDMPDir, "Cavalli_SHH_ageacatadj_dmp_MethFullResults.csv")
cavWNT_path <- file.path(cavDMPDir, "Cavalli_WNT_ageacatadj_dmp_MethFullResults.csv")
cavG3_path <- file.path(cavDMPDir, "Cavalli_Group3_ageacatadj_dmp_MethFullResults.csv")
cavG4_path <- file.path(cavDMPDir, "Cavalli_Group4_ageacatadj_dmp_MethFullResults.csv")
newALL_path <- file.path(newDMPDir, "Newcastle_agecatadj_dmp_MethFullResults.csv")
newSHH_path <- file.path(newDMPDir, "Newcastle_SHH_ageacatadj_dmp_MethFullResults.csv")
newWNT_path <- file.path(newDMPDir, "Newcastle_WNT_ageacatadj_dmp_MethFullResults.csv")
newG3_path <- file.path(newDMPDir, "Newcastle_Group3_ageacatadj_dmp_MethFullResults.csv")
newG4_path <- file.path(newDMPDir, "Newcastle_Group4_ageacatadj_dmp_MethFullResults.csv")

# Set the directory where the processed data will be kept
dir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results"
cavDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Cavalli"
newDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Newcastle"

# File paths to the DMP files for each subgroup
cavALL_path <- file.path(cavDMPDir, "Cavalli_agecatadj_dmp_MethFullResults.csv")
cavSHH_path <- file.path(cavDMPDir, "Cavalli_SHH_ageacatadj_dmp_MethFullResults.csv")
cavWNT_path <- file.path(cavDMPDir, "Cavalli_WNT_ageacatadj_dmp_MethFullResults.csv")
cavG3_path <- file.path(cavDMPDir, "Cavalli_Group3_ageacatadj_dmp_MethFullResults.csv")
cavG4_path <- file.path(cavDMPDir, "Cavalli_Group4_ageacatadj_dmp_MethFullResults.csv")
newALL_path <- file.path(newDMPDir, "Newcastle_agecatadj_dmp_MethFullResults.csv")
newSHH_path <- file.path(newDMPDir, "Newcastle_SHH_ageacatadj_dmp_MethFullResults.csv")
newWNT_path <- file.path(newDMPDir, "Newcastle_WNT_ageacatadj_dmp_MethFullResults.csv")
newG3_path <- file.path(newDMPDir, "Newcastle_Group3_ageacatadj_dmp_MethFullResults.csv")
newG4_path <- file.path(newDMPDir, "Newcastle_Group4_ageacatadj_dmp_MethFullResults.csv")

# # Set the directory where the counts files are found
# counts_dir <- here::here("analyses", "mutation_counts")
# 
# # File paths to the counts files for LGAT and medulloblastoma
# lgat_path <- file.path(counts_dir, "LGAT_gene-mutations.tsv")
# mb_path <- file.path(counts_dir, "Medulloblastoma_gene-mutations.tsv")
# 
# # Reference directory & FLAGS (frequently mutated genes)
# ref_dir <- here::here("data", "reference")
# flags_path <- file.path(ref_dir, "FLAGS.txt")
```

Outputs
```{r outputs}
# set the directory to save output plots
results_dir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/CNVplots"

# create the results directory if it does not exist
if (!dir.exists(results_dir)){
  dir.create(results_dir, recursive = TRUE)
}
```

### Download Data
Data previously downloaded

```{r downloads}

```

### Define functions
No functions right now

```{r functions}

```

### Reading data

We'll first read in the DMP files data for Cavalli, Newcastle, and healthy brain as tables. For the upset plot, I will have to use the gene names rather than the 450K probeID because different arrays were used. So I clean up the gene names in the DMP files that have duplicates in entries separated by ; because of multiple transcripts. I will also remove any duplicate genes due to probes associated to the same gene. Finally, I'll filter the healthy brain data to those DMPs with adjusted q-val < 0.05 and also removed duplicate gene names.

```{r read_data}

d <- c(cavSHH_path, cavWNT_path,cavG3_path,cavG4_path,newSHH_path, newWNT_path,newG3_path,newG4_path)
readCounts <- function(name){
  print(name)
  tmp <- read_csv(name)
  # colnames(tmp)[7] <- "counts"
  # colnames(tmp)[1] <- "locID"
  #tmp <- filter(tmp, Chr %in% c(1:22, "X","Y"))
  group <- str_remove(name, "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/")
  print(group)
  group <- str_remove(group, "_ageacatadj_dmp_MethFullResults.csv") 
  print(group)
  tmp$group <- group
  # tmp <- tmp[,c("locID","counts","sampleID")]
  tmp
}
all <- map_dfr(d, readCounts)

all <- left_join(all, ann450k_df)
all$start <- all$pos-1
all <- all %>% 
  separate(group, sep="/", c("cohort","group")) %>%
  separate(group, sep="_", c("cohortrep","subgroup"))
all_bed <- all[,c("chr","start","pos","probeID","cohort","subgroup")]
all_bed <- filter(all_bed, cohort=="Cavalli")
# write_tsv(all_bed,paste0(results_dir,"/","CavDMPloc.bed"))
# for (sub in unique(all_bed$subgroup)) {
#   dmp <- filter(all_bed, subgroup==sub)
#   write.table(dmp, file =paste0(results_dir,"/",sub,"_dmp.bed"), col.names=T, row.names=F, quote=F, sep = "\t")
# }

# read in intersections to plot CNV from

d <- list.files(path=results_dir, pattern = ".CNVfeature_intersect.txt")
readCNV <- function(name){
  tmp <- read_tsv(paste0(results_dir,"/",name, sep=""))
  colnames(tmp) <- c("chrom","start","end","probeID","DMPcohort","subgroup","CNV_chrom","CNV_start","CNV_end","ID","seg.mean","cohort","subgroup","pval","seg.median","overlap")
  tmp
}
all_intersect <- map_dfr(d, readCNV)
#all_intersect_bed <- all[,c("chrom","loc.start","loc.end","ID","seg.mean","cohort","subgroup","pval","seg.median")]
#write.table(all_bed, file = "CNVcombined.bed", col.names = T, row.names = F, quote = F, sep = "\t")





```

Read in pheno files
```{r readpheno}
phenoNew <- read_csv(paste0(phenoDir,"/processed/Newcastle/Newcastle_pheno_v3.csv"))
phenoCav <- read_csv(paste0(phenoDir,"/pheno_CavalliCombined_subgrouped.csv"))
phenoCav$sex <- 
  factor(phenoCav$predictedSex, 
         levels=c("Female","Male"),
         labels=c("F", # Reference
                  "M"))
pheno <- phenoNew[c("geo_accession","sample","sex")]
phenoC <- phenoCav[,c("geo_accession_Meth","Study_ID","sex")]
colnames(phenoC) <- c("geo_accession","sample","sex")
pheno <- rbind(pheno,phenoC)
all_intersect_pheno <- left_join(all_intersect, pheno,by=c("ID"="geo_accession"))

```

## Create boxplots of CNV by sex

```{r boxplot, out.width="100%"}
bp <- ggplot(subset(all_intersect_pheno,!is.na(sex) & subgroup...6=="Group3"), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot(add = "jitter")
G3_bp <- bp + facet_wrap(~ probeID, scales="free_y") + 
  stat_compare_means(aes(group=sex, ), hide.ns=T, size=2,label.x=1.5) +
  theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) +
  theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + 
  ggtitle("Medulloblastoma Group3 CNV segment mean comparisons at DMPs by sex")
G3_bp

#+ geom_rect(data = subset(all_intersect_pheno, !is.na(sex) & subgroup...6=="Group4" & probeID %in% c("cg09067967","cg11643285","cg02758552","cg17238319")), fill = NA, colour = "red", xmin = -Inf,xmax = Inf, ymin = -Inf,ymax = Inf, aes(size=5))

bp <- ggplot(subset(all_intersect_pheno,!is.na(sex) & subgroup...6=="Group4"), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot(add = "jitter")
G4_bp <- bp + facet_wrap(~ probeID, scales="free_y") + 
  stat_compare_means(aes(group=sex, ), hide.ns=T, size=2, label.x=1.5) +
  theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) +
  theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + 
  ggtitle("Medulloblastoma Group4 CNV segment mean comparisons at DMPs by sex")
G4_bp
#+ geom_rect(data = subset(all_intersect_pheno, !is.na(sex) & subgroup...6=="Group4" & probeID %in% c("cg22345911")),fill = NA, colour = "red", xmin = -Inf,xmax = Inf,ymin = -Inf,ymax = Inf, aes(size=5))

bp <- ggplot(subset(all_intersect_pheno,!is.na(sex) & subgroup...6=="WNT"), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot(add = "jitter")
WNT_bp <- bp + facet_wrap(~ probeID, scales="free_y") + 
  stat_compare_means(aes(group=sex, ), hide.ns=T, size=2, label.x=1.5) +
  theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) +
  theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + 
  ggtitle("Medulloblastoma WNT CNV segment mean comparisons at DMPs by sex")
WNT_bp

SHH_intersect <- subset(all_intersect_pheno, !is.na(sex) & subgroup...6=="SHH")
SHH_DMPs <- unique(SHH_intersect$probeID)
SHH_DMPs_1 <- SHH_DMPs[1:32]
SHH_DMPs_2 <- SHH_DMPs[33:64]
SHH_DMPs_3 <- SHH_DMPs[65:96]
SHH_DMPs_4 <- SHH_DMPs[97:130]

bp <- ggplot(subset(SHH_intersect, probeID %in% SHH_DMPs_1), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot()
SHH_bp1 <- bp + facet_wrap(~ probeID, scales="free_y") + stat_compare_means(aes(group=sex, ), label="p.signif", hide.ns=T, size=6, label.x=1.25) + theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) + theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + ggtitle("Medulloblastoma SHH CNV segment mean comparisons at DMPs by sex (Part I)")
SHH_bp1

bp <- ggplot(subset(SHH_intersect, probeID %in% SHH_DMPs_2), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot()
SHH_bp2 <- bp + facet_wrap(~ probeID, scales="free_y") + stat_compare_means(aes(group=sex, ), hide.ns=T, size=2, label.x=1.25) + theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) + theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + ggtitle("Medulloblastoma SHH CNV segment mean comparisons at DMPs by sex (Part II)")
SHH_bp2

bp <- ggplot(subset(SHH_intersect, probeID %in% SHH_DMPs_3), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot()
SHH_bp3 <- bp + facet_wrap(~ probeID, scales="free_y") + stat_compare_means(aes(group=sex, ), hide.ns=T, size=2, label.x=1.25) + theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) + theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + ggtitle("Medulloblastoma SHH CNV segment mean comparisons at DMPs by sex (Part III)")
SHH_bp3

bp <- ggplot(subset(SHH_intersect, probeID %in% SHH_DMPs_4), aes(x=sex,y=seg.mean, fill=sex)) + geom_boxplot()
SHH_bp4 <- bp + facet_wrap(~ probeID, scales="free_y") + stat_compare_means(aes(group=sex, ), hide.ns=T, size=2, label.x=1.25) + theme_bw(base_size=14) + scale_fill_manual(values=c("red", "black")) + theme(axis.title.x=element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.2, hjust = 0.95)) + ggtitle("Medulloblastoma SHH CNV segment mean comparisons at DMPs by sex (Part IV)")
SHH_bp4
```





### Saving plots

```{r save_plots}
# png(file=file.path(results_dir, "UpsetPlot_compareDMPs_healthybrain.png"),
# width = 1200, height =750, units = 'px')
# p1
# dev.off()

# ggsave(
#   file.path(results_dir, "UpsetPlot_compareDMPs_healthybrain.png"),
#   p1,
# )

```


## Session information

We like to end all of our notebooks with a call to a function that provides a brief report about the versions of R and all packages that were used in the notebook.
This can be very useful for debugging down the line, as different versions of packages, R, or even the operating system can sometimes result in slightly or significantly differen outputs.

The most common such function is `sessionInfo()`, which is part of base R and is what we are using here, but a nice alternative is the function `sessioninfo::session_info()` (part of the `sessioninfo` package), which provides a somewhat more neatly-formatted report with a bit more information.

```{r session_info}
sessionInfo()
```
