---
title: "MB Sex Differences in DNA Methylation in subgroups - CNV boxplots by sex"
author: "Rachel Moss"
date: "Aug 29, 2022"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
    toc_float: yes
editor_options:
  chunk_output_type: inline
---

## Introduction

I will use UpSetR package to compare the Cavalli and Newcastle DMPs for each subgroup to the DMPs in healthy brain from  [GSE15745](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE15745)
in an upset plot. 

## Setup

### Load packages

```{r packages, message=FALSE, results=FALSE, warning=FALSE}
library(UpSetR)
library(readr)
library(tidyr)
library(dplyr)
library(ggplot2)
```

### Specify files and paths

These DMP files for each subgroup for Cavalli and Newcastle are stored on a shared [google drive](https://drive.google.com/drive/folders/1MtKqg-v7USwvqbHVqPo_RiIwcu5O9nkb?usp=sharing).
The healthy brain DMPs are also on a shared [google drive](https://drive.google.com/drive/folders/1so_iPCp3Q4MlN-iAuQtimCt5-IzDYpDs?usp=sharing)

```{r inputs}
# Set the directory where the processed data will be kept
cavDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Cavalli"
newDMPDir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/Newcastle"
brainDMP <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Healthy Brain /GSE15745/GSE15745.top.table.CRBLM_M-F_withCHR.tsv"

# File paths to the DMP files for each subgroup
cavALL_path <- file.path(cavDMPDir, "Cavalli_agecatadj_dmp_MethFullResults.csv")
cavSHH_path <- file.path(cavDMPDir, "Cavalli_SHH_ageacatadj_dmp_MethFullResults.csv")
cavWNT_path <- file.path(cavDMPDir, "Cavalli_WNT_ageacatadj_dmp_MethFullResults.csv")
cavG3_path <- file.path(cavDMPDir, "Cavalli_Group3_ageacatadj_dmp_MethFullResults.csv")
cavG4_path <- file.path(cavDMPDir, "Cavalli_Group4_ageacatadj_dmp_MethFullResults.csv")
newALL_path <- file.path(newDMPDir, "Newcastle_agecatadj_dmp_MethFullResults.csv")
newSHH_path <- file.path(newDMPDir, "Newcastle_SHH_ageacatadj_dmp_MethFullResults.csv")
newWNT_path <- file.path(newDMPDir, "Newcastle_WNT_ageacatadj_dmp_MethFullResults.csv")
newG3_path <- file.path(newDMPDir, "Newcastle_Group3_ageacatadj_dmp_MethFullResults.csv")
newG4_path <- file.path(newDMPDir, "Newcastle_Group4_ageacatadj_dmp_MethFullResults.csv")

# # Set the directory where the counts files are found
# counts_dir <- here::here("analyses", "mutation_counts")
# 
# # File paths to the counts files for LGAT and medulloblastoma
# lgat_path <- file.path(counts_dir, "LGAT_gene-mutations.tsv")
# mb_path <- file.path(counts_dir, "Medulloblastoma_gene-mutations.tsv")
# 
# # Reference directory & FLAGS (frequently mutated genes)
# ref_dir <- here::here("data", "reference")
# flags_path <- file.path(ref_dir, "FLAGS.txt")
```

Outputs
```{r outputs}
# set the directory to save output plots
results_dir <- "/Users/moss0134/Google Drive/Research/SexDiffMedullo/Analysis_RMM_new/Results/compareDMPs"

# create the results directory if it does not exist
if (!dir.exists(results_dir)){
  dir.create(results_dir, recursive = TRUE)
}
```

### Download Data
Data previously downloaded

```{r downloads}

```

### Define functions
No functions right now

```{r functions}

```

### Reading data

We'll first read in the DMP files data for Cavalli, Newcastle, and healthy brain as tables. For the upset plot, I will have to use the gene names rather than the 450K probeID because different arrays were used. So I clean up the gene names in the DMP files that have duplicates in entries separated by ; because of multiple transcripts. I will also remove any duplicate genes due to probes associated to the same gene. Finally, I'll filter the healthy brain data to those DMPs with adjusted q-val < 0.05 and also removed duplicate gene names.

```{r read_data}
cavALL <- read_csv(cavALL_path)
cavSHH <- read_csv(cavSHH_path)
cavWNT <- read_csv(cavWNT_path)
cavG3 <- read_csv(cavG3_path)
cavG4 <- read_csv(cavG4_path)
newALL <- read_csv(newALL_path)
newSHH <- read_csv(newSHH_path)
newWNT <- read_csv(newWNT_path)
newG3 <- read_csv(newG3_path)
newG4 <- read_csv(newG4_path)
brain <- read_tsv(brainDMP)

cavALL <- cavALL %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
cavSHH <- cavSHH %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
cavWNT <- cavWNT %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
cavG3 <- cavG3 %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
cavG4 <- cavG4 %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
newALL <- newALL %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
newSHH <- newSHH %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
newWNT <- newWNT %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
newG3 <- newG3 %>% separate_rows(UCSC_RefGene_Name) %>% distinct()
newG4 <- newG4 %>% separate_rows(UCSC_RefGene_Name) %>% distinct()

cALL_L <- unique(cavALL$UCSC_RefGene_Name)
cSHH_L <- unique(cavSHH$UCSC_RefGene_Name)
cWNT_L <- unique(cavWNT$UCSC_RefGene_Name)
cG3_L <- unique(cavG3$UCSC_RefGene_Name)
cG4_L <- unique(cavG4$UCSC_RefGene_Name)
nALL_L <- unique(newALL$UCSC_RefGene_Name)
nSHH_L <- unique(newSHH$UCSC_RefGene_Name)
nWNT_L <- unique(newWNT$UCSC_RefGene_Name)
nG3_L <- unique(newG3$UCSC_RefGene_Name)
nG4_L <- unique(newG4$UCSC_RefGene_Name)

brain_adj <- filter(brain, adj.P.Val<0.05)
brain_L <- unique(brain_adj$Gene.symbol)

brain_Lt <- data.frame("symbol"=brain_L,"brain"=1)
cSHH_Lt <- data.frame("symbol"=cSHH_L,"Cav SHH"=1)
cWNT_Lt <- data.frame("symbol"=cWNT_L,"Cav WNT"=1)
cG3_Lt <- data.frame("symbol"=cG3_L,"Cav Group3"=1)
cG4_Lt <- data.frame("symbol"=cG4_L,"Cav Group4"=1)
nSHH_Lt <- data.frame("symbol"=nSHH_L,"New SHH"=1)
nWNT_Lt <- data.frame("symbol"=nWNT_L,"New WNT"=1)
nG3_Lt <- data.frame("symbol"=nG3_L,"New Group3"=1)
nG4_Lt <- data.frame("symbol"=nG4_L,"New Group4"=1)

listI <- list(brain_Lt,cSHH_Lt,cWNT_Lt,cG3_Lt,cG4_Lt,nSHH_Lt,nWNT_Lt,nG3_Lt,nG4_Lt) %>% purrr::reduce(full_join, by = "symbol")
# remove NA gene found in the brain data
listI <- listI[-6,]
# replace all NAs with 0 in the table
listI[is.na(listI)] <- 0

```

## Create upset plot
Now I create my upset plot using these lists.
```{r upset, out.width="100%"}
listInput1 <- list("Healthy Brain" = brain_L, "Cavalli SHH" = cSHH_L, "Cavalli WNT" = cWNT_L, "Cavalli Group3" = cG3_L, "Cavalli Group4" = cG4_L, "Newcastle SHH" = nSHH_L, "Newcastle WNT" = nWNT_L, "Newcastle Group3" = nG3_L, "Newcastle Group4" = nG4_L)

listInput2 <- list("Healthy Brain" = brain_L, "Cavalli ALL" = cALL_L, "Cavalli SHH" = cSHH_L, "Cavalli WNT" = cWNT_L, "Cavalli Group3" = cG3_L, "Cavalli Group4" = cG4_L, "Newcastle ALL" = nALL_L, "Newcastle SHH" = nSHH_L, "Newcastle WNT" = nWNT_L, "Newcastle Group3" = nG3_L, "Newcastle Group4" = nG4_L)

p1 <- upset(fromList(listInput1), sets=c("Newcastle Group4","Cavalli Group4","Newcastle Group3","Cavalli Group3","Newcastle WNT","Cavalli WNT","Newcastle SHH","Cavalli SHH", "Healthy Brain"), order.by="degree", keep.order=TRUE, mb.ratio = c(0.55, 0.45),text.scale = c(1.3, 1.3, 1, 1, 2, 1.3))
p1

p2 <- upset(fromList(listInput2), sets=c("Newcastle Group4","Cavalli Group4","Newcastle Group3","Cavalli Group3","Newcastle WNT","Cavalli WNT","Newcastle SHH","Cavalli SHH", "Newcastle ALL","Cavalli ALL","Healthy Brain"), order.by="degree", keep.order=TRUE)


```


### Print all intersections

```{r intersections}
data_with_intersection <- listI %>%
  unite(col = "intersection", -c("symbol"), sep = "")
data_with_intersection %>%
  group_by(intersection) %>%
  summarise(n = n()) %>%
  arrange(desc(n))
intersections <- data_with_intersection %>%
  group_by(intersection) %>%
  summarise(list = list(symbol)) %>%
  mutate(list = setNames(list, intersection)) %>%
  pull(list)

get_intersect_members <- function (x, ...){
  require(dplyr)
  require(tibble)
  x <- x[,sapply(x, is.numeric)][,0<=colMeans(x[,sapply(x, is.numeric)],na.rm=T) & colMeans(x[,sapply(x, is.numeric)],na.rm=T)<=1]
  n <- names(x)
  x %>% tibble::rownames_to_column() -> x
  l <- c(...)
  a <- intersect(names(x), l)
  ar <- vector('list',length(n)+1)
  ar[[1]] <- x
  i=2
  for (item in n) {
    if (item %in% a){
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '>= 1')
        i <- i + 1
      }
    } else {
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '== 0')
        i <- i + 1
      }
    }
  }
  do.call(filter, ar) %>% tibble::column_to_rownames() -> x
  return(x)
}

```


### Saving plots

```{r save_plots}
# png(file=file.path(results_dir, "UpsetPlot_compareDMPs_healthybrain.png"),
# width = 1200, height =750, units = 'px')
# p1
# dev.off()

# ggsave(
#   file.path(results_dir, "UpsetPlot_compareDMPs_healthybrain.png"),
#   p1,
# )

```


## Session information

We like to end all of our notebooks with a call to a function that provides a brief report about the versions of R and all packages that were used in the notebook.
This can be very useful for debugging down the line, as different versions of packages, R, or even the operating system can sometimes result in slightly or significantly differen outputs.

The most common such function is `sessionInfo()`, which is part of base R and is what we are using here, but a nice alternative is the function `sessioninfo::session_info()` (part of the `sessioninfo` package), which provides a somewhat more neatly-formatted report with a bit more information.

```{r session_info}
sessionInfo()
```
